### PR TITLE
Mon 21239 new notification test--draft--

### PR DESCRIPTION
**Pull request by sfarouq-ext**
Mon 21239 new notification test

## Type of change

- [ x] Patch fixing an issue (non-breaking change)
- [x ] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [ ] 22.04.x
- [ ] 22.10.x
- [ ] 23.04.x
- [ ] 23.10.x
- [ x] 24.04.x (master)

<h2> How this pull request can be tested ? </h2>

Please describe the **procedure** to verify that the goal of the PR is matched. Provide clear instructions so that it can be **correctly tested**.

Any **relevant details** of the configuration to perform the test should be added.

## Checklist

- [x ] I have followed the **coding style guidelines** provided by Centreon
- [x ] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [ x] I have commented my code, especially **hard-to-understand areas** of the PR.
- [x ] I have made corresponding changes to the **documentation**.
- [x ] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: tests/resources/Engine.py [hunk 1]
```diff
@@ -647,6 +714,9 @@ def build_configs(self, hosts: int, services_by_host: int, debug_level=0):
             with open(f"{config_dir}/escalations.cfg", "w") as f:
                 pass
 
+            f = open(config_dir + "/meta_services.cfg", "w")
+            f.close()
+
```
Comment (bouda1) [inline tests/resources/Engine.py:719]: Do we really need to create meta_services.cfg?

File: tests/resources/Engine.py [hunk 2]
```diff
@@ -952,6 +1022,16 @@ def ctn_engine_config_set_value_in_escalations(idx: int, desc: str, key: str, va
     with open(f"{ETC_ROOT}/centreon-engine/config{idx}/escalations.cfg", "w") as ff:
         ff.writelines(lines)
 
+def ctn_engine_config_set_value_in_dependencies(idx: int, desc: str, key: str, value: str):
+    with open(f"{ETC_ROOT}/centreon-engine/config{idx}/dependencies.cfg", "r") as ff:
```
Comment (bouda1) [inline tests/resources/Engine.py:1026]: Please add the function description

File: tests/resources/Engine.py [hunk 3]
```diff
@@ -2167,8 +2246,19 @@ def ctn_create_escalations_file(poller: int, name: int, SG: str, contactgroup: s
...
+    engine.create_dependencies_file(poller, dependenthost, host, dependentservice, service)
+
+def ctn_create_dependenciesgrp_file(poller: int, dependentservicegroup: str, servicegroup: str):
+    engine.create_dependenciesgrp_file(poller, dependentservicegroup, servicegroup)
+
+def ctn_create_dependencieshst_file(poller: int, dependenthost: str, host: str):
+    engine.create_dependencieshst_file(poller, dependenthost, host)
 
+def ctn_create_dependencieshstgrp_file(poller: int, dependenthostgrp: str, hostgrp: str):
+    engine.create_dependencieshstgrp_file(poller, dependenthostgrp, hostgrp)
```
Comment (bouda1) [inline tests/resources/Engine.py:2261]: All these new functions don't have any description

File: tests/resources/Engine.py [hunk 4]
```diff
@@ -2228,14 +2317,14 @@ def ctn_engine_config_remove_tag(poller: int, tag_id: int):
         else:
             tag_begin_idx = tag_begin_idx + 1
 
-    with open(filename, "w") as f:
-        f.writelines(lines)
+    f = open(filename, "w")
+    f.writelines(lines)
+    f.close()
```
Comment (bouda1) [inline tests/resources/Engine.py:2322]: The old version was really better

File: tests/resources/Engine.py [hunk 5]
```diff
@@ -2181,8 +2271,7 @@ def ctn_create_template_file(poller: int, typ: str, what: str, ids: list):
         what (str): A string. It depends on what type of template.
         ids (list): For each integer in this list, a template is defined.
     """
-    engine.ctn_create_template_file(poller, typ, what, ids)
-
+    engine.create_template_file(poller, typ, what, ids)
```
Comment (bouda1) [inline tests/resources/Engine.py:2274]: is it normal you remove the ctn_ prefix?

File: tests/resources/Engine.py [hunk 6]
```diff
@@ -3159,3 +3248,23 @@ def ctn_get_service_command(host_id: int, service_id: int):
         logger.console(
             f"Unable to find the command id of service ({host_id};{service_id})")
         return None
+
+def ctn_set_service_state(id: int, state: int):
```
Comment (bouda1) [inline tests/resources/Engine.py:3252]: the description function has a standard. Look at others functions for examples.

Comment (bouda1) [reply tests/resources/Engine.py:3252]: I'm almost sure there was already a function to do this job...

File: tests/resources/db_conf.py [hunk 7]
```diff
@@ -123,15 +123,15 @@ def create_conf_db(self):
                 for i in range(1, self.hosts_count + 1):
                     name = f"checkh{i}"
                     cursor.execute(
-                        f"INSERT INTO command (command_name,command_line) VALUES (\"{name}\",\"{ENGINE_HOME}/check.pl --id 0 --state {i}\")")
+                        "INSERT INTO command (command_name,command_line) VALUES (\"{2}\",\"{0}/check.pl --id 0 --state {1}\")".format(ENGINE_HOME, i, name))
```
Comment (bouda1) [inline tests/resources/db_conf.py:126]: Here again, you remove a good code to replace it by a bad one.

File: tests/resources/db_conf.py [hunk 8]
```diff
@@ -123,15 +123,15 @@ def create_conf_db(self):
...
                     self.command[name] = cursor.lastrowid
                 connection.commit()
 
                 # Insertion of SERVICES COMMANDS
                 for i in range(1, self.commands_per_poller_count * self.instances + 1):
                     name = f"command_{i}"
                     cursor.execute(
-                        f"INSERT INTO command (command_name,command_line) VALUES (\"{name}\",\"{ENGINE_HOME}/check.pl --id {i}\")")
+                        "INSERT INTO command (command_name,command_line) VALUES (\"{2}\",\"{0}/check.pl --id {1}\")".format(ENGINE_HOME, i, name))
```
Comment (bouda1) [inline tests/resources/db_conf.py:134]: again

File: tests/resources/engine-scripts/check.pl [hunk 9]
```diff
@@ -4,6 +4,10 @@
 use warnings;
 use Getopt::Long;
 
+if ($#ARGV < 1) {
+    die "The script must be used with one integer argument\n";
+}
```
Comment (bouda1) [inline tests/resources/engine-scripts/check.pl:9]: this is no more useful as we use real arguments with a parser for that.

**Review by bouda1 — CHANGES_REQUESTED** [review]

File: tests/broker-engine/external-commands2.robot [hunk 10]
```diff
@@ -1388,7 +1388,7 @@ BESERVCHECK
     ...    host_1
     ...    service_1
     ...    30
-    ...    ${VarRoot}/lib/centreon-engine/check.pl --id ${command_id}
+    ...    ${VarRoot}/lib/centreon-engine/check.pl --id ${command_param}
```
[the remark this answers is not in the document]
Comment (sfarouq-ext) [reply tests/broker-engine/external-commands2.robot:1391]: yes command_id was the correct one

File: tests/resources/Engine.py (hunk 1 above)
Comment (sfarouq-ext) [reply tests/resources/Engine.py:719]: no so i removed it

File: tests/resources/Engine.py (hunk 2 above)
Comment (sfarouq-ext) [reply tests/resources/Engine.py:1026]: done

File: tests/resources/Engine.py (hunk 3 above)
Comment (sfarouq-ext) [reply tests/resources/Engine.py:2261]: done

File: tests/resources/Engine.py (hunk 5 above)
Comment (sfarouq-ext) [reply tests/resources/Engine.py:2274]: yes cause it refers to the staticmethod 
@staticmethod
    def create_template_file(poller: int, typ: str, what: str, ids):

File: tests/resources/Engine.py (hunk 6 above)
Comment (sfarouq-ext) [reply tests/resources/Engine.py:3252]: yes the tests pass without it so i remove it

File: tests/resources/db_conf.py (hunk 7 above)
Comment (sfarouq-ext) [reply tests/resources/db_conf.py:126]: done

File: tests/resources/db_conf.py (hunk 8 above)
Comment (sfarouq-ext) [reply tests/resources/db_conf.py:134]: done

File: tests/resources/engine-scripts/check.pl (hunk 9 above)
Comment (sfarouq-ext) [reply tests/resources/engine-scripts/check.pl:9]: i ve removed it

File: tests/broker-engine/notifications.robot [hunk 11]
```diff
@@ -0,0 +1,1303 @@
...
+    Ctn Engine Config Set Value In Contacts    0    John_Doe    service_notification_commands    command_notif
+
+    ${start}    Get Current Date
+    Ctn Start Broker
+    Ctn Start Engine
+
+    # Let's wait for the external command check start
+    ${content}    Create List    check_for_external_commands()
+    ${result}    Ctn Find In Log With Timeout    ${engineLog0}    ${start}    ${content}    60
+    Should Be True    ${result}    check_for_external_commands() should be available.
```
Comment (bouda1) [inline tests/broker-engine/notifications.robot:126]: Ctn Wait For Engine To Be Ready...

File: tests/broker-engine/notifications.robot [hunk 12]
```diff
@@ -0,0 +1,1303 @@
...
+    # It's time to schedule a downtime
+    Ctn Schedule Service Fixed Downtime    host_1    service_1    3600
+
+    ${result}    Ctn Check Service Downtime With Timeout    host_1    service_1    1    60
+    Should Be True    ${result}    service must be in downtime
+
+    Ctn Delete Service Downtime    host_1    service_1
+
+    Ctn Process Service Result Hard    host_1    service_1    ${2}    The service_1 is CRITICAL
+
```
Comment (bouda1) [inline tests/broker-engine/notifications.robot:137]: This test is not useful.
* You create a service
* You put a downtime on it.
* you remove the downtime
* you set the service state to CRITICAL HARD.
* And then you check a notification is well sent.

Why the downtime would have an impact on the service if it is removed?

An interesting test would be:
* Create the service
* Put a downtime on it.
* Set its state to CRITICAL HARD.
* Check no notification is sent
* Remove the downtime
* Check there is a notification
* Set the state to OK
* Check there is a recovery notification.

**Review by bouda1 — CHANGES_REQUESTED** [review]

File: tests/broker-engine/notifications.robot [hunk 13]
```diff
@@ -0,0 +1,1303 @@
...
+    Ctn Engine Config Set Value In Contacts    0    John_Doe    service_notification_commands    command_notif
+
+    ${start}    Get Current Date
+    Ctn Start Broker
+    Ctn Start Engine
+
+    # Let's wait for the external command check start
+    ${content}    Create List    check_for_external_commands()
+    ${result}    Ctn Find In Log With Timeout    ${engineLog0}    ${start}    ${content}    60
+    Should Be True    ${result}    check_for_external_commands() should be available.
```
[the remark this answers is not in the document]
Comment (sfarouq-ext) [reply tests/broker-engine/notifications.robot:46]: done

**Pull request closed without merging**